lib/ecto/repo/preloader.ex

```diff
@@ -176,23 +178,57 @@ defmodule Ecto.Repo.Preloader do
         struct, _counts -> raise ArgumentError, "expected #{inspect(struct)} to contain embed `#{field}`"
       end)
 
-    embed_structs = preload_each(embed_structs, repo_name, sub_preloads, tuplet)
-    structs = load_embeds(card, field, structs, embed_structs, Enum.reverse(counts), [])
+    # It is not possible for an embed to preload through a join
+    # Therefore, we don't consider associations coming from queries
+    embed_structs = preload_each(embed_structs, repo_name, sub_preloads, [], tuplet)
+    structs = put_through_or_embed(card, field, structs, embed_structs, Enum.reverse(counts), [])
     preload_embeds(structs, embeds, repo_name, tuplet)
   end
 
-  defp load_embeds(_card, _field, [], [], [], acc), do: Enum.reverse(acc)
+  defp preload_throughs(structs, [] = _throughs, _, _, _), do: structs
 
-  defp load_embeds(card, field, [struct | structs], embed_structs, [0 | counts], acc),
-    do: load_embeds(card, field, structs, embed_structs, counts, [struct | acc])
+  defp preload_throughs(
+         structs,
+         [{_, _, false = _from_query?} | throughs],
+         repo_name,
+         query_assocs,
+         tuplet
+       ) do
+    # We only preload through associations directly if they were previously loaded by a query.
+    # Otherwise, we preload the association chain.
```

Comment:
```suggestion
    # We only preload through associations directly if they were previously loaded by a query.
    # Otherwise, we preload the association chain.
```

Can we clarify here? Because we have been struggling with nomenclature, so being more concrete will help. Let's use `Ecto.Query.preload` and `Ecto.Repo.preload`?